### PR TITLE
Refactor by extracting inbound and outbound adapter logic

### PR DESCRIPTION
**Pull request by warrenzhu25**

### CODE REVIEW

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/AbstractAzureMessageHandler.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    protected MessageConverter messageConverter;
+    private boolean sync = false;
+    private ListenableFutureCallback<Void> sendCallback;
+    private EvaluationContext evaluationContext;
+    private Expression sendTimeoutExpression = new ValueExpression<>(DEFAULT_SEND_TIMEOUT);
+    private Expression partitionKeyExpression;
+
+    public AbstractAzureMessageHandler(String destination, SendOperation<D> sendOperation) {
+        Assert.hasText(destination, "destination can't be null or empty");
+        Assert.notNull(sendOperation, "sendOperation can't be null");
```
Comment (Incarnation-p-lee) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/AbstractAzureMessageHandler.java]: How about `@NonNull` from springframework ?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Memorize function execution result
+ *
+ * @author Warren Zhu
+ */
+public class Memoizer {
```
Comment (sophiaso) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java]: `Memoizer` typo? And the method names below too.

Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java]: I followed from https://opencredo.com/lambda-memoization-in-java-8/ and https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/concurrent/Memoizer.html

Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java]: public final class Memoizer?

Comment (yaweiw) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java]: also add a private constructor to this class

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Tuple.java [hunk 3]
```diff
@@ -12,11 +12,23 @@
     private final T first;
     private final U second;
 
-    public Tuple(T first, U second) {
+    private Tuple(T first, U second) {
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Tuple.java]: interesting to see Java doesn't have Tuple implemented in SDK. Wanted to bring it up for discussion: from design perspective, do you think this is better https://msdn.microsoft.com/en-us/library/system.tuple(v=vs.110).aspx? e.g. make it static, have a create method, no getFirst() method etc.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java [hunk 4]
```diff
@@ -71,62 +69,72 @@ public void setCheckpointStorageAccountContainer(String checkpointStorageAccount
...
-
-            try {
-                return EventHubClient
-                        .createSync(getOrCreateConnectionString(eventHubName), Executors.newSingleThreadExecutor());
-            } catch (EventHubException | IOException e) {
-                throw new EventHubRuntimeException("Error when creating event hub client", e);
-            }
-        });
+    public Function<String, EventHubClient> getEventHubClientCreator() {
+        return Memoizer.memoize(clientsByName, this::createEventHubClient);
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java]: can this:: be removed?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java [hunk 5]
```diff
@@ -34,30 +37,25 @@
  */
 public class DefaultEventHubClientFactory implements EventHubClientFactory, DisposableBean {
     private static final Log LOGGER = LogFactory.getLog(DefaultEventHubClientFactory.class);
-    // eventHubName -> eventHubClient
-    private final ConcurrentHashMap<String, EventHubClient> clientMap = new ConcurrentHashMap<>();
 
-    // eventHubName -> connectionString
-    private final ConcurrentHashMap<String, String> connectionStringMap = new ConcurrentHashMap<>();
+    private final Map<String, EventHubClient> clientsByName = new ConcurrentHashMap<>();
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java]: it's a 1 to 1 mapping so i guess clientByName better?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/AbstractAzureMessageHandler.java (hunk 1 above)
Comment (yaweiw) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/AbstractAzureMessageHandler.java]: i think we should be consistent across all our Java code base e.g. annotation, getter setter etc.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java (hunk 2 above)
Comment (Incarnation-p-lee) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Memoizer.java]: try `@NoArgsConstructor(access = AccessLevel.PRIVATE)` here.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java [hunk 6]
```diff
@@ -57,12 +57,13 @@ public EventHubTemplate(EventHubClientFactory clientFactory) {
         Assert.hasText(eventHubName, "eventHubName can't be null or empty");
         Assert.notNull(message, "message can't be null");
         try {
-            EventHubClient client = this.clientFactory.getOrCreateEventHubClient(eventHubName);
+            EventHubClient client = this.clientFactory.getEventHubClientCreator().apply(eventHubName);
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java]: Wondering why we'd introduce this Memoizer and Function? I indeed think clientFactory.getOrCreateEventHubClient is better than clientFactory.getEventHubClientCreator().apply.

Comment (yaweiw) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java]: Memoization is used to cache expensive resources. EventHubClient isn't cacheable as it is tightly coupled with connectionstring and eventhub name. A new eventHubName means a new creation.
https://www.javatips.net/api/commons-lang-master/src/main/java/org/apache/commons/lang3/concurrent/Memoizer.java

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/AbstractAzureMessageHandler.java (hunk 1 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/AbstractAzureMessageHandler.java]: Agree.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java (hunk 4 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java]: no. It's method reference syntax.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java (hunk 5 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactory.java]: Yes.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java (hunk 6 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java]: 1. Memoizer is used to remove duplication of `putIfAbsent` operation of map. Another benefit is to separate the concern of different function by taking adavantage of functional composition.
2. EventHubClient is only coupled by eventhub name. So we can use Memoizer to cache with name as input and client as output.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Tuple.java (hunk 3 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/Tuple.java]: I have static factory method `of`.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java (hunk 6 above)
Comment (yaweiw) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java]: Good learning to me. Thanks.

**Pull request merged**